src/gui/sdlmain.cpp

```diff
@@ -4228,8 +4229,8 @@ void config_add_sdl() {
 
 	pstring->Set_help(
 	        "Set the host video driver's vertical synchronization (vsync) mode:\n"
-	        "  auto:      Limit vsync to beneficial cases, such as when using an\n"
-	        "             interpolating VRR display in fullscreen (default).\n"
+	        "  auto:      Limit vsync to beneficial cases, such as when in Windowed-mode\n"
```

Comment:
No, vsync in windowed mode just happens on all modern OSes. It's impossible to get tearing in windowed mode in any modern OS, no matter how hard you try. So the previous comment and behaviour was correct.

Reply:
For the record, compositor-level vsync was introduced in Windows Vista, so in 2007. You could disable compositor level vsync in Win 8 (somehow), but it's that's no longer an option in Win 10+.

Reply:
That's why I'm defaulting to assuming vsync is enabled in windowed mode (only when `vsync = auto`) instead of possibly getting it wrong through analyzing benchmark results if the timing or frameates just happen to run a bit faster than normal. 
In other words, it's the most reasonable default in windowed mode (because compositors are already enforcing vsync). 

The consequence of getting it wrong (that is, the drivers and SDL telling you that vsync is off when it's actually enforced) is that 70 Hz frames rates will stutter and "breathe" in waves that go above and below the 60 Hz range.

Reply:
The presentation mode needs to know if we can present frames faster or slower than the DOS rate. 

On a 60hz display in windowed mode, the answer, 99% of the time, we'll be orcefully blocked if we try to present faster more 60 fps. 

And yet, querying both SDL and the OpenGL might tell you "vsync is off, the sky is the limit". But yet, our micro benchmark will actually prove that we're clamped at 60 fps (which is our last chance to detect I'd we're actually vsync'd).

Reply:
You're absolutely right! I take it back, as per my other comment. I thought you setting this flag would mean we in DOSBox are actively enforcing vsync in windowed mode by asking SDL, or asking the OpenGL driver directly.

If it's just passing this implicit info about windowed mode to your benchmarker, then yeah, we're in 100% agreement, and this is most definitely what we should do! Apologies 😄

Reply:
Although... it's okay that we're making this assumption for the _benchmarker_, but this help setting talks about what we enforce on the driver level:

> Set the host video driver's vertical synchronization (vsync) mode:

So actually no, we're not setting vsync ourselves in windowed mode; we _assume_ vsync in on in windowed mode, right? Based on what you wrote above, I think that is what's happening, so the help text is actually incorrect now, no?

Reply:
>  No, vsync in windowed mode just happens on all modern OSes.

I mean, that is just false on a Linux system.  I've written enough simple games in SDL to know you get thousands of FPS and a very warm computer if you don't enable VSync, even in Windowed mode.

Maybe the compositioner helps tearing but it does not cap the FPS in the way that real VSync does.  You're also not even guaranteed to have a compositioner running.  KDE, for example, has a feature that turns off its compositioner on demand when it detects an SDL application or game running in order to boost performance.

Reply:
> If it's just passing this implicit info about windowed mode to your benchmarker, 

Need to double-check 😅! I'm positive the info is passed through, however need to check if we're trying to enable it as well.

The dilemma, at least on Linux and the Pi, is that everything can lie. So basically it's all just best effort. 
I've had cases where SDL and OpenGL say vsync is enabled, get 0==success return codes from those C-calls.. but nope: the compositor or drive can override you. (And vice-versa). 
This is what drove the need for the microbenchmark. 

But yeah. . Will double check it tomorrow and worstcase, will just make auto the assumed state.

Reply:
> > No, vsync in windowed mode just happens on all modern OSes.
> 
> I mean, that is just false on a Linux system. I've written enough simple games in SDL to know you get thousands of FPS and a very warm computer if you don't enable VSync, even in Windowed mode.
> Maybe the compositioner helps tearing but it does not cap the FPS in the way that real VSync does.

Actually, that's true. Maybe I should read the whole code again sometime... But ok, we kinda need to enable OpenGL level vsync level anyway to act as back-pressure to avoid generating too many frames. That's actually too. Mea culpa. I'm tired, didn't have much sleep... 😩 

> You're also not even guaranteed to have a compositioner running. KDE, for example, has a feature that turns off its compositioner on demand when it detects an SDL application or game running in order to boost performance.

Yeah, Linux might be special, but it's been enforced on macOS and Windows for the last 10+ years.

> EDIT: And sorry, I will not accept your Reddit thread with 3 upvotes as proof 😆

Haha, fair enough 😄 "Bbbb... but I found it on the internet!...." 😅

Reply:
1) X compositors can be suspended or not be present at all (which will lead to tearing), and compositor-side vsync can be disabled (which, again, will lead to tearing). In either case, actual framerate is not limited, it's up to the program how many frames it renders (rendering is separate to presenting).
2) Wayland compositors recently (yes, it took years) got a protocol to enable tearing. But that only allows tearing presentation, it does not affect rendering speed in any way. Even without that protocol, the framerate is not limited.
3) Vulkan has mailbox presentation mode. Unlimited framerate, but vsynced presentation. Works the same on any platform that supports it, - that includes both Windows and macOS. While Staging currently does not implement Vulkan-based renderer, I brought it up for a better understanding of what compositors actually do.

Now, enabling vsync both on program and compositor side might actually lead to additional latency (remember, we're talking frames, not milliseconds) and even dropped frames. But OpenGL has absolutely no way of integrating with any platform's compositor, it isn't aware of such things. On Windows, we'd have to directly talk to either DWM or DXGI to get properly timed vsync, for example. (Same goes for Vulkan, really.) Only exclusive fullscreen allows the program to properly drive presentation, and that's discouraged on Windows and outright does not exist on Linux.

Reply:
Thanks @GranMinigun. So what default do you suggest?

(Btw, I started thinking about how the hell does lagless vsync even work in WinUAE. Well I know it needs fullscreen mode, it cannot be used in windowed mode (the option is even greyed out). Then in fullscreen I'm not even using exclusive fullscreen with the DirectX 12 renderer, well because it does not exist on Windows 10. But it does brute-force buffer swapping _with intentional tearing_ to achieve 1/N frame latencies where N is the frame slice number. It works splendidly... I just don't understand how is that possible... Anyway, side remark.)

Reply:
Another remark, on Windows 10 I can achieve perfect butter-smooth scrolling in games like Pinball Dreams only one way: to use my custom tweaked 70 Hz mode on my monitor, then CFR with vsync enabled. Vsync is crucial otherwise it tends to get out of sync a couple of times per second and jumps a bit. So it does make a difference, but only if I can hit the exact DOS rate (or very close).

Reply:
Hard to say. Windowed mode might behave differently in Wayland compositors that allow direct scan-out (so presentation might tear if vsync is disabled on program's side). In all other cases, it doesn't really matter, as presentation will be driven by the compositor (or lack of one) anyway. Probably leave it off, if we're limited by the emulated monitor's refresh rate anyway? We shouldn't care how often display memory contents change, only how often emulated monitor hits vblank.

WinUAE's sync method works because DX's presentation is handled by DXGI, so it's completely integrated with the compositor. That was the case ever since DX10, I believe. And it's possible to ask for the current scanline.

Reply:
> We shouldn't care how often display memory contents change, only how often emulated monitor hits vblank.

That's what I suggested last year and that would solve this whole frame duping and missed frame business. But @kcgen explained VFR is necessary for some use cases still and it can conserve power for example when you're using a DOS IDE that rarely updates the screen (although, there is usually a blinking cursor anyway...) But pretty much all other emulators just do present on emulated vblank like you said and then you either enable vsync or you don't. Simple and works so that's still my preference, for the record 😄 Just present at 70 Hz in VGA modes, even if my monitor is in 50 Hz mode, then let the OS deal with it. Then for the vsync on case you really need a VRR monitor for DOS, or you can sort of of use fixed monitor refresh rates for specific modes like me, but that's a bit crap... (especially for demos that switch between custom refresh rates in tweaked modes every 10 seconds...)

Reply:
> Just present at 70 Hz in VGA modes, even if my monitor is in 50 Hz mode, then let the OS deal with it

If vsync is enforced either by the user at the driver level or frame rates are being limited by OS (or compositor), even if vsync is reported as off, our present-on-vblank calls will block for a lot more than 1ms.

This is actually what lead to @kklobe originally writing the frame-pacer. Apple M1-based laptops running external displasys hard-block our presentation calls for > 10ms, caused brutal stalls and lag.

The frame-pacer was then rejoiced for boosting performance, because the same stalling was being experienced under Linux on the Pi; and other users also reported performance boosts (as did I, only Linux + 60 Hz system).

The downside of the frame-pacer was that it dropped the /next/ frame, which might have been important (which lead to more changes, like the presentation-mode throttling-and re-present approach on the front-side), which the pacer itself is still  a valuable failsafe on the backside.

Reply:
The root cause is that the presentation and emulation are running serially. If we can split them into separate threads, then the suggestion:

> Just present at 70 Hz in VGA modes, even if my monitor is in 50 Hz mode, then let the OS deal with it

Would be perfect - because the presentation layer can jam up against against any vsync blocks, while the emalator carry on :-)

Reply:
(updated in above comment)

Reply:
> If vsync is enforced either by the user at the driver level or frame rates are being limited by OS (or compositor), even if vsync is reported as off, our present-on-vblank calls will block for a lot more than 1ms.

Ok @kcgen but that's how other emulators work. If I run VICE to emulate a PAL 50 Hz C64 or WinUAE to emulate a PAL Amiga on a fixed 60 Hz display with vsync on, things will go comically wrong. Audio stutters, general lagginess and jerkiness, etc.; it will become 100% unusable. Solution? Don't do that unless you can provide a 50 Hz host rate to vsync to 😄 

In those emulators vsync on is a power user thing for people who know what they're doing. Everybody else; please use the vsync off default 😅 And in windowed mode the vsync option is greyed out; they must just present the frame at timed intervals without any vsync enforcement on the _emulator_ side. There will be zero tearing that way due to the compositor level syncing, or Linux with compositor level vsync off there will be tearing, but that's how it is. So to be clear they _never_ block or time anything to host rates in the vsync on mode; they continuously present frames at 50 Hz in a non-blocking fashion then it's the driver's business.

You're saying that we got blocking with this approach when _not_ enforcing vsync in our program, so even if we do `glSwapInterval(0)` and present frames at precisely timed intervals only, running on our own timers? That's very surprising if that's the case, to put it mildly. There should be no blocking in our app in that scenario. Am I completely confused here @GranMinigun ?

Can't we just do the same?

Reply:
It will block for the amount of time required to render the image and send it to the presentation engine, but it won't wait for vblank. We don't do threaded/async rendering/presentation.

Reply:
There's lots of systems that have 60 Hz vsync enforced at the driver or compositor-levels, and these user suffer badly when running dosbox playing fast 70 Hz games. Lots of slight micro jamups! I know.. I'm one of those people 😂

It also happens on kklobe's Apple laptop + external display.

I'm running a standard Linux + KDE, have it configured for the lowest latency, and OG DOSBox (as well as pure VFR mode) will give me only 45 fps in Quake 320x200 because of these jamups.

It happens to SBC users too running on the framebuffer, like in retropie. 

I believe the default Windows Intel settings are vsynced as well, as I had to dig around to figure out how to disable it. But some Intel driver vartianta are even worse and don't provide that option.

Reply:
Thanks for the explanation @kcgen (again and again 😅) Don't you think it would be worthwhile to start a documentation page on these presentation mode details? I can create a blank page template for you, then you can put in some rough info in there and I can help with rewording and stripping it down to easily digestible information, etc. if you just wanna do it quickly. Frankly, I could do it myself and use your above comments as a starting point, just wanted to ask you first. If busy or don't have time, I can give it a shot and you'll review it eventually anyway.

It's just all this info is super useful and necessary to explain the various presentation mode options to our users and I'll forget 50%+ of it by next week 😅 JustWorks(tm) by default is the best, no doubt, but we still need to explain things in the manual, and power users need to know. The release notes on presentation modes don't mention any of this background context, so going by that very few people could correctly guess why this is needed, why is it useful, and so on (well, hence I was super confused by the beginning; the release notes was the only info I was going by).

Btw, yeah, when emulating pre SVGA cards, fixed host rate setting could work 100% well. But starting with VGA, 640x480 is already 60 Hz, then with SVGA and VESA anything goes because by that time multisync monitors became the standard. So for perfect SVGA/VESA emulation, especially to run all those demoscene prod without throttling, a full VRR setup is non-negotiable. Just making this observation, there's nothing we can do about this 😄 But I guess it underlines the importance of throttled VFR, if we have to do VRF because of the above reasons.

Reply:
Yes, I can try to summarize it for the manual and power users; happy to fill in those details 👍